archive/imo/imo2014_q1.lean

```diff
@@ -0,0 +1,170 @@
+Authors: Adrián Doña Mateo
+-/
+
+import algebra.big_operators
+import data.pnat.basic
+
+/-!
+# IMO 2014 Q1
+
+Let `a₀ < a₁ < a₂ < ⋯` be an infinite sequence of positive integers.
+Prove that there exists a unique `n ≥ 1` such that
+
+  `aₙ < (a₀ + a₁ + ⋯ + aₙ) / n < aₙ₊₁`.
+
+This solutions is a translation of the official solution, which may be found as the
+solution to problem A1 [here](https://www.imo-official.org/problems/IMO2014SL.pdf).
+-/
+
+open_locale big_operators
+
+noncomputable theory
+
+/-- The integer sequence. -/
+variable (a : ℕ → ℤ)
+
+/-- We define an auxiliary sequence `d a : ℕ+ → ℤ` by
+  `d a n = (a 0 + a 1 + ⋯ + a n) - n * (a n)`. -/
+def d (n : ℕ+) : ℤ := (∑ i : fin (n + 1), a i) - n * (a n)
+
+lemma first_ineq_iff {n : ℕ+} :
+  (a n : ℚ) < (↑∑ i : fin (n + 1), a i : ℚ) / n ↔ 0 < d a n :=
+show _ < _ / ↑(n : ℤ) ↔ _, begin
+  have : (↑(a n) : ℚ) = ↑(a n) / ↑(1 : ℤ) := by simp,
+  rw [this, rat.div_lt_div_iff_mul_lt_mul (int.zero_lt_one)],
+  { simp [d, mul_comm] }, simp,
+end
+
+lemma second_ineq_iff {n : ℕ+} :
+  (↑∑ i : fin (n + 1), a i : ℚ) / n ≤ a (n + 1) ↔ d a (n + 1) ≤ 0 :=
+show _ / ↑(n : ℤ) ≤ _ ↔ _, begin
+  have : (↑(a (n + 1)) : ℚ) = ↑(a (n + 1)) / ↑(1 : ℤ) := by simp, rw this,
+  simp only [rat.div_num_denom, mul_one, rat.num_one, one_mul, rat.coe_int_denom, int.coe_nat_zero,
+    rat.coe_int_num, int.coe_nat_succ, zero_add, coe_coe],
+  rw rat.le_def, swap,
+  { norm_cast, have : 1 ≤ ↑n := pnat.one_le n, linarith }, swap,
+  { exact zero_lt_one },
+  simp only [d, @fin.sum_univ_cast_succ _ _ ↑(n + 1), mul_one, fin.coe_last, pnat.one_coe,
+    fin.coe_cast_succ, int.coe_nat_succ, pnat.add_coe, sub_nonpos, coe_coe, add_mul, one_mul],
+  rw [add_le_add_iff_right,	mul_comm], refl,
+end
+
+/-- We rephrase the original question into a question about `d a`. -/
+lemma ineq_iff {n : ℕ+} :
+  (a n : ℚ) < (↑∑ i : fin (n + 1), a i : ℚ) / n ∧
+  (↑∑ i : fin (n + 1), a i : ℚ) / n ≤ a (n + 1) ↔
+  0 < d a n ∧ d a (n + 1) ≤ 0 :=
+⟨λ h, ⟨(first_ineq_iff a).1 h.1, (second_ineq_iff a).1 h.2⟩,
+  λ h, ⟨(first_ineq_iff a).2 h.1, (second_ineq_iff a).2 h.2⟩⟩
+
+lemma d_one : d a 1 = a 0 :=
+show a 0 + (a 1 + 0) - 1 * a 1 = a 0, by simp
+
+/-- If `a` is strictly increasing, then `d a` is strictly decreasing. -/
+lemma ddes (hinc : ∀ n, a n < a (n + 1)) (n : ℕ+) : d a (n + 1) < d a n :=
+lt_of_sub_neg $
+  calc ((∑ i : fin (n + 2), a i) - (n + 1) * a (n + 1))
+        - ((∑ i : fin (n + 1), a i) - n * a n)
+      = ((∑ i : fin (n + 1), a i) + a (n + 1) - (n + 1) * a (n + 1))
+        - ((∑ i : fin (n + 1), a i) - n * a n) : by simp [fin.sum_univ_cast_succ]
+  ... = n * (a n - a (n + 1))                  : by ring
+  ... < 0
+    : mul_neg_iff.mpr (or.inl ⟨by simp, sub_neg_of_lt (hinc n)⟩)
+
+section descending
+
+/-- In this section we prove that for any strictly descending sequence `f : ℕ+ → ℤ`
+  and integer `x < f m` for some `m`, there is a unique index `n ≥ m` such that
+  `f (n + 1) ≤ x < f n`. -/
+
```

Comment:
This whole section feels like it should be in mathlib proper, suitably generalised. (e.g. you could generalise away from `int` with the hypotheses that `f` tends to infinity, which then you get for free from strictness in the `int` case)

Presumably in any actual IMO solution this fact would not even receive a proof; assuming that's the case, I think it's a good indication the material doesn't belong in `archive`.

Reply:
It's also just a distraction in the proof here that the domain is `ℕ+`.

Reply:
I'm thinking about how to generalize this. Let's flip the inequality around for simplicity. We have a strict monotone function `f : α → β` from a `succ_order α` to a `preorder β`. To prove our intended result, the quickest approach is to take the greatest `n` such that `f n < x`, and prove that `x ≤ f (succ n)`. Doing this requires a `linear_order β`, but more crucially, it requires us to take maximums of bounded sets in `α`. This seems like a really strong condition, and I'm not sure how to state it.

Reply:
Alternatively, we could take the least `n` such that `x ≤ f n`, and prove that `f (pred n) < x`. In either case, I don't really think our result will generalize much further than for `α = ℤ`.

Reply:
I believe this is the correct generalization:
```lean
theorem unique_cross_lt_le {α β : Type*} [linear_order α] [succ_order α] [preorder β] {f : α → β}
  (hf : strict_mono f) (b : β) : {a | f a < b ∧ b ≤ f (succ_order.succ a)}.subsingleton :=
begin
  rintros a₁ ⟨_, _⟩ a₂ ⟨_, _⟩,
  apply le_antisymm;
  refine succ_order.le_of_lt_succ (hf.lt_iff_lt.1 (lt_of_lt_of_le _ _)),
  assumption'
end
```
I'll try to PR this along with related results, and I'll simplify the proof by using this.